### PR TITLE
ref: normalize file references

### PR DESCRIPTION
**Pull request by qinxuye**
## Summary

- render user attachments as inline file chips that open the right-side preview, without duplicate raw Markdown links or image cards
- show file-type-specific chip icons using MIME metadata first and filename extensions as a fallback
- centralize file reference parsing and generation around canonical `file:<id>` references while accepting legacy `file://<id>` values
- route workspace tools, media tools, Telegram, and Feishu through the shared file reference handling

## Root cause

Connector uploads could store legacy `file://<id>` references. The chat UI rendered the runtime Markdown reference and attachment metadata independently, producing duplicate attachment UI. Backend tools also normalized file references in different places, so `understand_images` treated the legacy reference as a path until an absolute path was supplied.

## Impact

Uploaded photos and files now appear as a single inline, clickable file chip with the corresponding image, audio, video, document, spreadsheet, presentation, archive, or code icon. Existing messages remain compatible, and media tools can resolve both canonical and legacy file ID references without interpreting real `file:///absolute/path` URIs as IDs.

## Validation

- 255 targeted backend tests passed
- 53 targeted frontend tests passed
- frontend type checking passed
- Python and frontend lint checks passed
- commit hooks passed, including ruff, mypy, npm lint, and npm type-check

### CODE REVIEW

File: frontend/src/components/chat/user-message-content.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+export const stripAttachedFileRefs = (
+  message: string,
+  files: UserMessageFile[]
+): string => {
+  const attachedIds = new Set(
+    files
+      .map((file) => file.file_id?.trim())
+      .filter((fileId): fileId is string => Boolean(fileId))
+  )
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply frontend/src/components/chat/user-message-content.tsx:67]: Addressed in 84e6326c. Added a shared sanitizer with array, object, and field-type validation, and stripAttachedFileRefs now only reads sanitized entries. Regression coverage includes null, primitive, and malformed object values.

File: frontend/src/components/chat/user-message-content.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  message,
+  files,
+  onPreview,
+}: UserMessageContentProps) {
+  const displayMessage = stripAttachedFileRefs(message, files)
+
+  return (
+    <div className="whitespace-pre-wrap max-h-60 overflow-y-auto">
+      {displayMessage}
+      {files.map((file, index) => {
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply frontend/src/components/chat/user-message-content.tsx]: Addressed in 84e6326c. UserMessageContent sanitizes attachments before rendering or preview dispatch, so only validated files reach the JSX map and callback. Added a component regression test with malformed entries.

File: src/xagent/core/file_ref.py [hunk 3]
```diff
@@ -24,6 +24,52 @@
...
+    if value is None:
+        return None
+
+    raw = str(value).strip()
+    if not raw.startswith("file:"):
+        return None
+
+    parsed = urlsplit(raw)
+    if parsed.scheme != "file" or parsed.query or parsed.fragment:
+        return None
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/file_ref.py]: Addressed in 84e6326c. urlsplit is now wrapped in a ValueError fail-safe, with a regression test for a malformed bracketed authority.

File: src/xagent/core/file_ref.py [hunk 4]
```diff
@@ -24,6 +24,55 @@
...
+
+
+def build_file_id_ref(file_id: str) -> str:
+    """Return the canonical internal URI for a registered file id."""
+    normalized = str(file_id).strip()
+    if not normalized or normalized in {".", ".."}:
+        raise ValueError("file_id must be a non-empty identifier")
+    if "/" in normalized or "\\" in normalized:
+        raise ValueError("file_id must not contain path separators")
+    return f"file:{quote(normalized, safe='')}"
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/file_ref.py:76]: Addressed in 2df20321. build_file_id_ref now validates both the original and percent-decoded identifier for path separators and encoded dot segments, with regression coverage for %2F, %5C, and %2E%2E.

File: tests/core/test_file_ref.py
Comment (rogercloud) [inline tests/core/test_file_ref.py:1]: `parse_file_id_ref` has untested branches for `None`, `""`, `"file:"` (bare empty id), and `"file://"` (bare empty authority); `build_file_id_ref`'s encode→decode round-trip and the canonical `file:<id>` vision-tool path are also uncovered — consider adding parametrized cases for these.

File: frontend/src/components/file/file-type-icon.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  if (mime.includes("presentation") || mime.includes("powerpoint")) {
+    return "presentation"
+  }
+  if (/\b(zip|rar|7z|tar|gzip|bzip2)\b/.test(mime)) return "archive"
+  if (
+    mime.includes("json") ||
+    mime.includes("xml") ||
+    mime.includes("javascript")
+  ) {
+    return "code"
```
Comment (rogercloud) [inline frontend/src/components/file/file-type-icon.tsx:120]: This generic `document` check must stay below the `spreadsheet`/`presentation` checks above — xlsx/pptx MIME strings contain the substring "document", so reordering would misclassify them; please add a comment marking the ordering as intentional and a test that feeds a real xlsx/pptx MIME string.

**Review by rogercloud — APPROVED** [review]
## Summary

This PR centralizes file-reference handling around a single canonical form (`file:<id>`, with legacy `file://<id>` still accepted) by introducing two helpers — `parse_file_id_ref` / `build_file_id_ref` in `src/xagent/core/file_ref.py` — and routing workspace tools, media tools, Telegram, Feishu, and the frontend chat/build components through them. This collapses three previously-scattered ad-hoc `file:` / `file://` string-slicing sites into one place and fixes the observed symptoms: connector uploads storing legacy `file://<id>` references, and the chat UI rendering runtime Markdown references and attachment metadata independently (producing duplicate raw links / image cards). On the frontend, user attachments now render as file-type-aware inline chips (MIME-first icon selection, filename-extension fallback) that drive the right-side preview.

## Re-review update

This is a re-review. All 5 open items raised across the 2 prior gemini-code-assist rounds were independently re-verified against current HEAD (`2df20321`) by fresh reads of the code plus passing tests — not merely trusted from the author's replies — and all are confirmed **FIXED**:

1. **FIXED** — `stripAttachedFileRefs` null / non-object array element crash: `sanitizeUserMessageFiles` now filters malformed entries before any `.file_id` access; regression test with `[null, "invalid", {}, image]` passes.
2. **FIXED** — `UserMessageContent` render + `onPreview` callback both now operate on sanitized files; full-component "ignores malformed attachment entries" regression test passes.
3. **FIXED** — `urlsplit` unhandled `ValueError` in `parse_file_id_ref`: wrapped in `try/except ValueError: return None`; `("file://[invalid", None)` case passes.
4. **FIXED (was HIGH)** — `stripAttachedFileRefs` crash on null / undefined / non-string `message`: signature now takes `message: unknown` and guards with `typeof message !== "string"` (stricter than the originally suggested fix); tests assert both `null` and object inputs return `""` without throwing. (No explicit author "addressed in commit X" reply exists for this one, but the fix is present and tested regardless.)
5. **FIXED (security)** — `build_file_id_ref` percent-encoded path traversal (`%2F`, `%5C`, `%2E%2E`): now unquotes and checks both raw and decoded forms for separators and dot-segments; regression tests for all 3 variants pass. The read/parse side rejects the same encodings, so there is no write-side/read-side asymmetry.

## Design verdict: acceptable-with-reservations

The core design is sound and a genuine root-cause fix. Security-sensitive disambiguation (`file:<id>` / `file://<id>` are IDs; `file:///abs/path` and `file://host/share/path` are NOT) is implemented correctly via `urlsplit` + netloc/path checks + separator rejection, and path containment stays enforced downstream (traced, no traversal regression). `audio_tool.py` / `image_tool.py` correctly drop their inline `file:` stripping and delegate to `workspace.resolve_path_with_search` — the right leverage rather than more scatter.

**Scope note (not a blocker):** commit `afca3eac` bundles an unrelated change — deleting `src/xagent/skills/builtin/poster-design/SKILL.md` (253 lines) and pruning the now-dangling poster-design references from `src/xagent/skills/selector.py`'s system prompt. Verified as clean mechanical dead-code removal (zero remaining `poster-design` references repo-wide; the `selector.py` diff is purely subtractive; dependent files `agent_tool.py`, `test_dag.py`, `test_utils.py` were updated consistently). It is unrelated to file-reference normalization, though — recommend splitting into its own PR for a clean history.

## Line-level findings

### `tests/core/test_file_ref.py` — [minor] test coverage gap
`parse_file_id_ref` is not exercised for `None`, `""`, `"file:"` (bare empty id), or `"file://"` (bare empty authority), though all four branches exist (`src/xagent/core/file_ref.py` ~L35-59). `build_file_id_ref`'s encode→decode round-trip through `parse_file_id_ref` is also untested, and `tests/core/tools/test_vision_tool.py` only drives `understand_images` with the legacy `file://<id>` form, never the canonical `file:<id>` end-to-end. **Suggestion:** add parametrized cases for the four empty/None inputs, a round-trip assertion, and one canonical-form vision-tool case.

### `frontend/src/components/file/file-type-icon.tsx` L98-129 (`getFileVisualKind`) — [minor] undocumented load-bearing MIME ordering
The `spreadsheet` (L107) and `presentation` (L110) substring checks must precede the generic `document` check (L119-127): the xlsx and pptx MIME strings (`application/vnd.openxmlformats-officedocument.spreadsheetml.sheet`, and pptx similarly) both contain the substring `"document"`, so reordering the generic branch above the specific ones would silently misclassify xlsx/pptx as generic documents. No comment records this invariant, and no test passes a real xlsx/pptx MIME string (the existing spreadsheet test only exercises the extension-fallback path). **Suggestion:** add a short comment noting the ordering is intentional, and a test that feeds an actual xlsx/pptx MIME string to catch a reordering regression.

## Simplification opportunities

The markdown file-reference regex `/\[([^\]]+)\]\(file:(?:\/\/)?([^)]+)\)/g` was touched in 4 places in this PR with byte-for-byte identical patterns (each changed from strict `file:\/\/` to the same optional-`//` form). Extract one shared `FILE_MARKDOWN_LINK_RE` constant and import it; safe and behavior-preserving (same 2-capture-group contract, same `.replace(regex, callback)` global-flag usage).

- `frontend/src/components/build/agent-builder.tsx` L324: [dup-regex] duplicate file-link regex literal. Import shared `FILE_MARKDOWN_LINK_RE`.
- `frontend/src/components/build/agent-builder.tsx` L438: [dup-regex] duplicate file-link regex literal. Import shared `FILE_MARKDOWN_LINK_RE`.
- `frontend/src/components/chat/ChatInput.tsx` L730: [dup-regex] duplicate file-link regex literal. Import shared `FILE_MARKDOWN_LINK_RE`.
- `frontend/src/components/chat/ChatMessage.tsx` L168: [dup-regex] duplicate file-link regex literal. Import shared `FILE_MARKDOWN_LINK_RE`.

net: -6 to -9 lines possible

Note: the 5th, related pattern in `frontend/src/components/chat/user-message-content.tsx` L23 (`/!?\[[^\]\n]*\]\(file:(?:\/\/)?([^)\n]+)\)/g`) has a different contract (single capture group, `!?` image-marker prefix, newline exclusion) and should stay separate — do not fold it into the shared constant.

## Test results

- Backend: `pytest tests/core/test_file_ref.py tests/core/test_workspace_path_containment.py tests/core/tools/test_vision_tool.py tests/core/agent/test_dag.py tests/skills/test_utils.py` → **176 passed**.
- Frontend: `vitest run src/components/file/file-type-icon.test.tsx src/components/chat/user-message-content.test.tsx` → **11 passed** (2 files).

## Verdict

Approve with minor follow-ups. Design is sound, all 5 prior findings (including the HIGH null-message crash and the security percent-encoding traversal) are genuinely fixed and tested, and the full suite passes. The two new findings are minor test/robustness gaps; the regex de-duplication and the poster-design commit split are optional cleanliness improvements.

**Review by rogercloud — APPROVED** [review]
## Follow-up re-review — commit `d837bed3` ("fix: strengthen file reference coverage")

**What changed since the approved commit `2df20321`:** The single new commit `d837bed3` is test-and-comment only — no production logic changed. It adds 4 parametrized `parse_file_id_ref` cases (`None`, `""`, bare `"file:"`, bare `"file://"`) plus a `build_file_id_ref` → `parse_file_id_ref` round-trip assertion in `tests/core/test_file_ref.py`, extends `tests/core/tools/test_vision_tool.py` to cover the canonical `file:<id>` form end-to-end (previously only legacy `file://<id>` was exercised), and documents the OpenXML-before-generic-"document" MIME ordering invariant in `frontend/src/components/file/file-type-icon.tsx` with two new real-xlsx/pptx tests in `file-type-icon.test.tsx`. This directly closes both inline findings from the prior round.

This re-review also re-ran a full from-scratch Round 0 (design) + Round 1 (discovery) pass over the entire PR (all 27 changed files).

### Fresh Round 0 — Design verdict: **sound**

The `parse_file_id_ref`/`build_file_id_ref` pair in `src/xagent/core/file_ref.py` is a genuine root-cause fix. It correctly distinguishes canonical `file:<id>`, legacy `file://<id>` (id in authority), and real `file://`/`file:///abs/path` URIs — the latter are correctly rejected as ids, so workspace path-containment still applies to them. This collapses the previously-scattered ad-hoc stripping in `audio_tool.py`, `image_tool.py`, `workspace_file_tool.py`, and `telegram/utils.py` into one place, and the frontend markdown-ref regex was updated consistently at every call site (no straggler `file://`-only patterns remain). Backward compatibility with legacy refs is preserved and tested, and the security boundary (a real `file:///path` is never treated as an id) holds under test.

**Design-level concern (non-blocking, carried over from the prior round):** commit `afca3eac` bundles an unrelated change — deleting `src/xagent/skills/builtin/poster-design/SKILL.md` (253 lines) and pruning its references from `selector.py` — into this file-reference-normalization PR. Recommend splitting for clean history; not a functional defect.

### Prior-findings checklist (verified against HEAD `d837bed3`)

1. **FIXED** — `tests/core/test_file_ref.py` untested branches: all 4 `parse_file_id_ref` edge cases (`None`, `""`, bare `"file:"`, bare `"file://"`) are now covered, the round-trip assertion was added, and `test_vision_tool.py` now parametrizes over both canonical and legacy forms end-to-end.
2. **FIXED** — `frontend/src/components/file/file-type-icon.tsx`: MIME-ordering comment added at L107 ("Keep specific OpenXML MIME checks above the generic 'document' branch."), plus two tests feeding real xlsx (`application/vnd.openxmlformats-officedocument.spreadsheetml.sheet`) and pptx MIME strings that assert correct classification.
3. **STILL OPEN (advisory)** — the `poster-design` skill deletion in `afca3eac` (253 lines) is unrelated scope, still bundled here. Recommend splitting into its own PR. Non-blocking.
4. **CONFIRMED, still open** — the markdown file-reference regex remains duplicated (see Simplification opportunities below).

### New confirmed finding (fresh discovery)

**`src/xagent/core/tools/core/image_tool.py` / `audio_tool.py` — minor (low severity, non-blocking).** The diff removed the `file:`-prefix stripping from the no-workspace fallback branch of `_resolve_image_path` / `_resolve_audio_path` (the branch that runs when `self._workspace` is `None`, a supported configuration; these methods are called unconditionally from `edit_image`/`transcribe_audio`). With the strip gone, a raw `"file:<id>"` now falls through to `Path("file:<id>")` relative-path resolution, which always misses. This is **not a functional regression** — file_id→path resolution fundamentally depends on the workspace, so the old stripped id string wouldn't have resolved to a real file in the no-workspace case either — but the error path now carries a literal `file:` prefix into a doomed lookup, and no test covers the no-workspace + `file:`-prefix combination.
Suggested fix: restore the `parse_file_id_ref` normalization (or the prefix strip) at the top of the fallback branch for defensive consistency. Do not block merge on it.

### Simplification opportunities

- `agent-builder.tsx` L324 / L438, `ChatInput.tsx` L730, `ChatMessage.tsx` L168: [dedup] the file-reference regex `/\[([^\]]+)\]\(file:(?:\/\/)?([^)]+)\)/g` is duplicated byte-for-byte in 4 places. Extract a shared `FILE_MARKDOWN_LINK_RE` constant and reference it at all four sites. (The related but distinct regex in `user-message-content.tsx:23` has a different contract — single capture group, `!?` image-marker prefix, newline exclusion — and should NOT be folded in.)
- net: -6 to -9 lines possible

### Test results

Backend — `tests/core/test_file_ref.py`, `tests/core/test_workspace_path_containment.py`, `tests/core/tools/test_vision_tool.py`, `tests/core/agent/test_dag.py`, `tests/skills/test_utils.py`: 176+ passed. Frontend — `file-type-icon.test.tsx`, `user-message-content.test.tsx`: all passed.

### Verdict: **Approve** (approval carried forward)

The new commit is test/comment-only and resolves both prior inline findings; no new blocking issues. Remaining items are all minor/advisory: the unrelated `poster-design` deletion scope (recommend splitting), the regex dedup, and the defensive no-workspace `file:`-strip restoration — none of which block merge.
